src/python/pants/backend/python/lint/isort/rules.py

```diff
@@ -57,10 +63,48 @@ class Setup:
     original_digest: Digest
 
 
+def is_old_isort_version(version: Optional[str]) -> bool:
+    if version is None:
+        return False
+
+    requirement = cast(PackagingRequirement, Requirement.parse(version))
+    specifier = requirement.specifier
+
+    old_releases = (
+        *[f"4.3.{i}" for i in range(22)],
+        *[f"4.2.{i}" for i in range(16)],
+        *[f"4.1.{i}" for i in range(3)],
+        "4.0.0",
+    )
+    old_releases_match = any(old_release in specifier for old_release in old_releases)
+    if not old_releases_match:
+        return False  # if no old release matches, the version must be new
+
+    # if old releases match, the version may still be >=5 (e.g. isort!=5.0.0 or simply isort)
+    new_releases_match = any(  # This will match as soon as a new version matches
+        f"{major}.{minor}.{patch}" in specifier
+        for major in range(5, 10)
+        for minor in range(20)
+        for patch in range(50)
+    )
+    return not new_releases_match
+
+
 def generate_args(*, source_files: SourceFiles, isort: Isort, check_only: bool) -> Tuple[str, ...]:
-    # NB: isort auto-discovers config files. There is no way to hardcode them via command line
-    # flags. So long as the files are in the Pex's input files, isort will use the config.
+    # NB: isort auto-discovers config files in versions <5. Only for isort>=5 we need to hardcode them via command line
+    # flags.
     args = []
+
+    if isort.config:
+        # We don't really expect users of isort<4
+        if len(isort.config) >= 2:
+            logger.warning(
+                "Found more than one config file for isort, taking the last one (%s) since isort uses a single config"
+                "file: the last one provided. For more, customization, set [isort].args.",
+                isort.config[-1],
+            )
```

Comment:
You can pass `--settings-file` multiple times, although it looks like the order of the flags does matter:

```
❯ ./pants fmt src/python/pants/util/strutil_test.py --isort-config=build-support/flake8/.flake8 --isort-args='--settings-file=build-support/flake8/.flake8 --settings-file=pyproject.toml' --black-skip --docformatter-skip
10:38:59.77 [INFO] Completed: fmt - isort made no changes.
```

```
❯ ./pants fmt src/python/pants/util/strutil_test.py --isort-config=build-support/flake8/.flake8 --isort-args='--settings-file=pyproject.toml --settings-file=build-support/flake8/.flake8' --black-skip --docformatter-skip
10:38:10.14 [WARN] Completed: fmt - isort made changes.
Fixing src/python/pants/util/strutil_test.py
```

That second command overrides the pyproject.toml, which is not desirable.

It's not clear to me how Pants should handle this. That's pretty nuanced that users are going to need to set up `[isort].config` in the right order, and that's even further removed from how tricky it already is to get right using dedicated isort. 

--

Which makes me wonder, how are you feeling about this feature overall? isort continues to confuse me with how many different ways they have of loading config files, and their precedence rules are non-trivial. Pants has a constant tension of trying to have sensible defaults and automating things, but also not being "too clever".

Here, a major risk is that Pants auto-setting `--settings-path` means that there won't be a way to un-set that. The above advice to set `[isort].args` for more complex cases is bogus - that will not stop Pants from still setting the `argv` to use `--settings-path`. There's no way to turn off this feature if the `--settings-path` option is messing things up.

I'm leaning towards this being a misfeature, but am curious what you think?

Reply:
The Pants implementation works in the same way as isort. It simply picks the last config file - no merging going on, Pants only logs a warning to make users aware of this. I don't think that would make Pants too smart.
```bash
(ws) Cristian Matache@Cristian-XPS13 /d/workspace/python/src/isort_mess
$ ls -R
.:
non_std_location/  x.py  x.py_

./non_std_location:
pyproject1.toml  pyproject2.toml

(ws) Cristian Matache@Cristian-XPS13 /d/workspace/python/src/isort_mess
$ cat x.py
from typing import Union, Optional, List, Tuple, Dict, Sequence, Iterable, Type, TypeVar, overload, cast, Generic, AsyncIterable, Awaitable, AwaitableGenerator
(ws) Cristian Matache@Cristian-XPS13 /d/workspace/python/src/isort_mess
$ cat non_std_location/pyproject1.toml
[tool.isort]
line_length = 10
(ws) Cristian Matache@Cristian-XPS13 /d/workspace/python/src/isort_mess
$ cat non_std_location/pyproject2.toml
[tool.isort]
line_length = 120
multi_line_output = 2

(ws) Cristian Matache@Cristian-XPS13 /d/workspace/python/src/isort_mess
$ isort x.py --settings-path non_std_location/pyproject1.toml --settings-path non_std_location/pyproject2.toml
Fixing D:\workspace\python\src\isort_mess\x.py
(ws) Cristian Matache@Cristian-XPS13 /d/workspace/python/src/isort_mess
$ cat x.py
from typing import AsyncIterable, Awaitable, AwaitableGenerator, Dict, Generic, Iterable, List, Optional, Sequence, \
    Tuple, Type, TypeVar, Union, cast, overload

(ws) Cristian Matache@Cristian-XPS13 /d/workspace/python/src/isort_mess
$ cp x.py_ x.py
(ws) Cristian Matache@Cristian-XPS13 /d/workspace/python/src/isort_mess
$ cat x.py
from typing import Union, Optional, List, Tuple, Dict, Sequence, Iterable, Type, TypeVar, overload, cast, Generic, AsyncIterable, Awaitable, AwaitableGenerator

(ws) Cristian Matache@Cristian-XPS13 /d/workspace/python/src/isort_mess
$ isort x.py --settings-path non_std_location/pyproject2.toml --settings-path non_std_location/pyproject1.toml
Fixing D:\workspace\python\src\isort_mess\x.py
(ws) Cristian Matache@Cristian-XPS13 /d/workspace/python/src/isort_mess
$ cat x.py
from typing import (
    AsyncIterable,
    Awaitable,
    AwaitableGenerator,
    Dict,
    Generic,
    Iterable,
    List,
    Optional,
    Sequence,
    Tuple,
    Type,
    TypeVar,
    Union,
    cast,
    overload)
(ws) Cristian Matache@Cristian-XPS13 /d/workspace/python/src/isort_mess
$ isort --version

                 _                 _
                (_) ___  ___  _ __| |_
                | |/ _/ / _ \/ '__  _/
                | |\__ \/\_\/| |  | |_
                |_|\___/\___/\_/   \_/

      isort your imports, so you don't have to.

                    VERSION 5.6.4

(ws) Cristian Matache@Cristian-XPS13 /d/workspace/python/src/isort_mess
```

Reply:
Nested isort config files don't really make sense: 
```(ws) Cristian Matache@Cristian-XPS13 /d/workspace/python/src/isort_mess2/a
$ ls -R
.:
b/  pyproject.toml  x._py  x.py

./b:
__init__.py  pyproject.toml  y.py
(ws) Cristian Matache@Cristian-XPS13 /d/workspace/python/src/isort_mess2/a
$ cat x.py
from typing import Union, Optional, List, Tuple, Dict, Sequence, Iterable, Type, TypeVar, overload, cast, Generic, AsyncIterable, Awaitable, AwaitableGenerator
(ws) Cristian Matache@Cristian-XPS13 /d/workspace/python/src/isort_mess2/a
$ cat pyproject.toml
[tool.isort]
line_length = 120
multi_line_output = 2
(ws) Cristian Matache@Cristian-XPS13 /d/workspace/python/src/isort_mess2/a
$ cat b/y.py
from typing import Union, Optional, List, Tuple, Dict, Sequence, Iterable, Type, TypeVar, overload, cast, Generic, AsyncIterable, Awaitable, AwaitableGenerator
(ws) Cristian Matache@Cristian-XPS13 /d/workspace/python/src/isort_mess2/a
$ cat b/pyproject.toml
[tool.isort]
line_length = 10
(ws) Cristian Matache@Cristian-XPS13 /d/workspace/python/src/isort_mess2/a
$ isort .
Fixing D:\workspace\python\src\isort_mess2\a\x.py
Fixing D:\workspace\python\src\isort_mess2\a\b\y.py
(ws) Cristian Matache@Cristian-XPS13 /d/workspace/python/src/isort_mess2/a
$ cat x.py
from typing import AsyncIterable, Awaitable, AwaitableGenerator, Dict, Generic, Iterable, List, Optional, Sequence, \
    Tuple, Type, TypeVar, Union, cast, overload
(ws) Cristian Matache@Cristian-XPS13 /d/workspace/python/src/isort_mess2/a
$ cat b/y.py
from typing import AsyncIterable, Awaitable, AwaitableGenerator, Dict, Generic, Iterable, List, Optional, Sequence, \
    Tuple, Type, TypeVar, Union, cast, overload
(ws) Cristian Matache@Cristian-XPS13 /d/workspace/python/src/isort_mess2/a
$ cd b/
(ws) Cristian Matache@Cristian-XPS13 /d/workspace/python/src/isort_mess2/a/b
$ isort .
Fixing D:\workspace\python\src\isort_mess2\a\b\y.py
(ws) Cristian Matache@Cristian-XPS13 /d/workspace/python/src/isort_mess2/a/b
$ cat y.py
from typing import (
    AsyncIterable,
    Awaitable,
    AwaitableGenerator,
    Dict,
    Generic,
    Iterable,
    List,
    Optional,
    Sequence,
    Tuple,
    Type,
    TypeVar,
    Union,
    cast,
    overload)
```

Reply:
I am not sure if "parallel" isort configs make sense or if this pattern fits within the Pants mindset. 

```(ws) Cristian Matache@Cristian-XPS13 /d/workspace/python/src/isort_mess3
$ ls -R
.:
a/  b/

./a:
pyproject.toml  x._py  x.py

./b:
__init__.py  pyproject.toml  y.py
(ws) Cristian Matache@Cristian-XPS13 /d/workspace/python/src/isort_mess3
$ cat a/x.py
from typing import Union, Optional, List, Tuple, Dict, Sequence, Iterable, Type, TypeVar, overload, cast, Generic, AsyncIterable, Awaitable, AwaitableGenerator
(ws) Cristian Matache@Cristian-XPS13 /d/workspace/python/src/isort_mess3
$ cat a/pyproject.toml
[tool.isort]
line_length = 120
multi_line_output = 2
(ws) Cristian Matache@Cristian-XPS13 /d/workspace/python/src/isort_mess3
$ cat b/y.py
from typing import Union, Optional, List, Tuple, Dict, Sequence, Iterable, Type, TypeVar, overload, cast, Generic, AsyncIterable, Awaitable, AwaitableGenerator
(ws) Cristian Matache@Cristian-XPS13 /d/workspace/python/src/isort_mess3
$ cat b/pyproject.toml
[tool.isort]
line_length = 10
(ws) Cristian Matache@Cristian-XPS13 /d/workspace/python/src/isort_mess3
$ isort .
Fixing D:\workspace\python\src\isort_mess3\a\x.py
Fixing D:\workspace\python\src\isort_mess3\b\y.py
(ws) Cristian Matache@Cristian-XPS13 /d/workspace/python/src/isort_mess3
$ cat a/x.py
from typing import (AsyncIterable, Awaitable, AwaitableGenerator, Dict,
                    Generic, Iterable, List, Optional, Sequence, Tuple, Type,
                    TypeVar, Union, cast, overload)
(ws) Cristian Matache@Cristian-XPS13 /d/workspace/python/src/isort_mess3
$ cat b/y.py
from typing import (AsyncIterable, Awaitable, AwaitableGenerator, Dict,
                    Generic, Iterable, List, Optional, Sequence, Tuple, Type,
                    TypeVar, Union, cast, overload)
(ws) Cristian Matache@Cristian-XPS13 /d/workspace/python/src/isort_mess3
$ cd a/
(ws) Cristian Matache@Cristian-XPS13 /d/workspace/python/src/isort_mess3/a
$ isort .
Fixing D:\workspace\python\src\isort_mess3\a\x.py
(ws) Cristian Matache@Cristian-XPS13 /d/workspace/python/src/isort_mess3/a
$ cd ..
(ws) Cristian Matache@Cristian-XPS13 /d/workspace/python/src/isort_mess3
$ cd b/
(ws) Cristian Matache@Cristian-XPS13 /d/workspace/python/src/isort_mess3/b
$ isort .
Fixing D:\workspace\python\src\isort_mess3\b\y.py
(ws) Cristian Matache@Cristian-XPS13 /d/workspace/python/src/isort_mess3/b
$ cd ..
(ws) Cristian Matache@Cristian-XPS13 /d/workspace/python/src/isort_mess3
$ cat a/x.py
from typing import AsyncIterable, Awaitable, AwaitableGenerator, Dict, Generic, Iterable, List, Optional, Sequence, \
    Tuple, Type, TypeVar, Union, cast, overload
(ws) Cristian Matache@Cristian-XPS13 /d/workspace/python/src/isort_mess3
$ cat b/y.py
from typing import (
    AsyncIterable,
    Awaitable,
    AwaitableGenerator,
    Dict,
    Generic,
    Iterable,
    List,
    Optional,
    Sequence,
    Tuple,
    Type,
    TypeVar,
    Union,
    cast,
    overload)
```

Reply:
Discussed offline with @Eric-Arellano  and it looks like isort merges nested .isort.cfg files but does not merge pyproject.toml files. This can be closed.

Reply:
> Discussed offline with @Eric-Arellano

To recap a little, isort caring about the _ordering_ of `--settings-path` raised a lot of red flags. isort config file handling is finicky with their rules for the precedence of config files. 

This PR is excellent if you only have a single config file, but the risk for breaking users with multiple config files is too high.